#149: Moved the API documentation to this repository. (#150)
exasol/virtual-schema-common-java: doc/development/api/virtual_schema_api.md

    },
    "escapeChar": "%"
}
```

The same can be used for `predicate_like_regexp`.

Notes
* **escapeChar** is optional

##### BETWEEN

```json
{
    "type": "predicate_between",
    "expression": {
        ...
    },
    "left": {
        ...
    },
    "right": {
        ...
    }
}
```

##### IN

`<exp> IN (<const1>, <const2>)`

```json
{
    "type": "predicate_in_constlist",
    "expression": {
        ...
    }
    "arguments": [
        ...
    ]
}
```

##### IS JSON / IS NOT JSON

`exp1 IS JSON {VALUE | ARRAY | OBJECT | SCALAR} {WITH | WITHOUT} UNIQUE KEYS`
 (requires predicate capability `IS_JSON`)

```json
{
    "type": "predicate_is_json",
    "expression": {
        ...
    },
    "typeConstraint": "VALUE",
    "keyUniquenessConstraint": "WITHOUT UNIQUE KEYS"
}
```

Notes:

- typeConstraint is `"VALUE"`, `"ARRAY"`, `"OBJECT"`, or `"SCALAR"`.
- keyUniquenessConstraint is `"WITH UNIQUE KEYS"` or `"WITHOUT UNIQUE KEYS"`.

The same can be used for a predicate type `predicate_is_not_json` (requires predicate capability `IS_NOT_JSON`).

### Scalar Functions

Refer to the [Exasol Scalar Functions API Documentation](scalar_functions_api.md)

### Aggregate Functions

> Consistent with scalar functions. To be detailed: `star-operator`, `distinct`, ...

Author: I have no idea what it's supposed to mean

> Consistent with scalar functions. To be detailed: `star-operator`, `distinct`, ...

Reviewer: ```suggestion
```

{
    "type": "predicate_is_json",
    "expression": {
        ...
    },
    "typeConstraint": "VALUE",
    "keyUniquenessConstraint": "WITHOUT UNIQUE KEYS"
}
```

Notes:

- typeConstraint is `"VALUE"`, `"ARRAY"`, `"OBJECT"`, or `"SCALAR"`.
- keyUniquenessConstraint is `"WITH UNIQUE KEYS"` or `"WITHOUT UNIQUE KEYS"`.

The same can be used for a predicate type `predicate_is_not_json` (requires predicate capability `IS_NOT_JSON`).

### Scalar Functions

Refer to the [Exasol Scalar Functions API Documentation](scalar_functions_api.md)

### Aggregate Functions

#### Functions With a Common API

Consistent with scalar functions. To be detailed: `star-operator`, `distinct`, ...

An aggregate function with a single argument (consistent with multiple argument version):

```json
{
    "type": "function_aggregate",
    "name": "SUM",
    "arguments": [
    {
        ...
    }
    ]
}
```

An aggregate function with multiple arguments:

```json
{
    "type": "function_aggregate",
    "name": "CORR",
    "arguments": [
    {
        ...
    },
    {
        ...
    }
    ]
}
```

#### Special Cases of Aggregate Functions

This section contains functions that have a special API mapping.

> | Function Name       | API Mapping Link                                   |
> |---------------------|----------------------------------------------------|
> | AVG(exp)            | [AVG(exp) function](#avgexp)                       |
> | AVG(DISTINCT exp)   | [AVG(DISTINCT exp) function](#avgdistinct-exp)     |
> | COUNT(*)            | [COUNT(*) function](#count)                        |
> | COUNT(exp)          | [COUNT(exp) function](#countexp)                   |
> | COUNT(DISTINCT exp) | [COUNT(DISTINCT exp) function](#countdistinct-exp) |
> | COUNT((exp1, exp2)) | [COUNT((exp1, exp2)) function](#countexp1-exp2)    |
> | GROUP_CONCAT        | [GROUP_CONCAT function](#group_concat)             |

Reviewer: ```suggestion
| Function Name       | API Mapping Link                                   |
|---------------------|----------------------------------------------------|
| AVG(exp)            | [AVG(exp) function](#avgexp)                       |
| AVG(DISTINCT exp)   | [AVG(DISTINCT exp) function](#avgdistinct-exp)     |
| COUNT(*)            | [COUNT(*) function](#count)                        |
| COUNT(exp)          | [COUNT(exp) function](#countexp)                   |
| COUNT(DISTINCT exp) | [COUNT(DISTINCT exp) function](#countdistinct-exp) |
| GROUP_CONCAT        | [GROUP_CONCAT function](#group_concat)             |
```

#### Aggregate Functions Not Included in the API

> | Function Name   | Comment                                 |
> |-----------------|-----------------------------------------|
> | ANY             | The API uses the SOME function.         |
> | CORR            | Not included in the API.                |
> | COVAR_POP       | Not included in the API.                |
> | COVAR_SAMP      | Not included in the API.                |
> | GROUPING        | Not included in the API.                |
> | LISTAGG         | The API uses the GROUP_CONCAT function. |
> | PERCENTILE_CONT | Not included in the API.                |
> | PERCENTILE_DISC | Not included in the API.                |
> | REGR_*          | Not included in the API.                |

Reviewer: ```suggestion
| Function Name       | Comment                                 |
|---------------------|-----------------------------------------|
| ANY                 | The API uses the SOME function.         |
| CORR                | Not included in the API.                |
| COUNT((exp1, exp2)) | Not included in the API.                |
| COVAR_POP           | Not included in the API.                |
| COVAR_SAMP          | Not included in the API.                |
| GROUPING            | Not included in the API.                |
| LISTAGG             | The API uses the GROUP_CONCAT function. |
| PERCENTILE_CONT     | Not included in the API.                |
| PERCENTILE_DISC     | Not included in the API.                |
| REGR_*              | Not included in the API.                |
```

##### COUNT(exp)

`COUNT(exp)`
 (requires set-function capability `COUNT`)

```json
{
    "type": "function_aggregate",
    "name": "COUNT",
    "arguments": [
    {
        ...
    }
    ]
}
```
##### COUNT(*)

`COUNT(*)`

>  (requires set-function capability `COUNT` and `COUNT_STAR`)

Reviewer: ```suggestion
 (Requires set-function capability `COUNT_STAR`. Please notice, that the set-function capability `COUNT` is not required in this case.)
```

```json
{
    "type": "function_aggregate",
    "name": "COUNT"
}
```

##### COUNT(DISTINCT exp)

`COUNT(DISTINCT exp)`

>  (requires set-function capability `COUNT` and `COUNT_DISTINCT`)

Reviewer: ```suggestion
 (requires set-function capabilities `COUNT` and `COUNT_DISTINCT`)
```

```json
{
    "type": "function_aggregate",
    "name": "COUNT",
    "distinct": true,
    "arguments": [
    {
        ...
    }
    ]
}
```

> 
> ##### COUNT((exp1, exp2))
> 
> `COUNT((exp1, exp2))`
> (requires set-function capability `COUNT` and `COUNT_TUPLE`)
> 
> ```json
> {
>     "type": "function_aggregate",
>     "name": "COUNT",
>     "distinct": true,
>     "arguments": [
>     {
>         ...
>     },
>     {
>         ...
>     }
>     ]
> }
> ```

Reviewer: ```suggestion
```
Not implemented, yet.
```

##### AVG(exp)

`AVG(exp)`
 (requires set-function capability `AVG`)

```json
{
    "type": "function_aggregate",
    "name": "AVG",
    "arguments": [
    {
        ...
    }
    ]
}
```

##### AVG(DISTINCT exp)

`AVG(DISTINCT exp)`

>  (requires set-function capability `AVG` and `AVG_DISTINCT`)

Reviewer: ```suggestion
 (requires set-function capabilities `AVG` and `AVG_DISTINCT`)
```
